Cython/Compiler/Nodes.py

```diff
@@ -606,7 +606,15 @@ def analyse(self, base_type, env, nonempty=0, visibility=None, in_pxd=False):
             self.dimension = self.dimension.analyse_const_expression(env)
             if not self.dimension.type.is_int:
                 error(self.dimension.pos, "Array dimension not integer")
-            size = self.dimension.get_constant_c_result_code()
+            size = None
+            if self.dimension.type.is_const:
+                if self.dimension.entry.defined_in_pxd:
+                    error(self.pos, "Array dimension declared in pxd file")
```

Comment:
This doesn't seem to be tested. Could you explain (or add a comment) why this isn't allowed? It doesn't seem immediately wrong to me if a constant is defined in the module's `.pxd` file. Or is it only about cimported pxd files?

Reply:
> This doesn't seem to be tested.

Not yet. The PR is still not 100% ready for review.

Reply:
> Could you explain (or add a comment) why this isn't allowed? It doesn't seem immediately wrong to me if a constant is defined in the module's .pxd file.

This error is raised only when `cimport`ed `const` variable is used as array dimension. Array dimensions in C require compile time value, which is not the case of values `cimport`ed.  In `const` variables declared within module this can be circumvented via values declared as macros - see https://github.com/cython/cython/pull/5242#issuecomment-1712574908

Reply:
> This doesn't seem to be tested.

Test added.